{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1537ad46",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c7d2c746",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate an ID mapping dictionary for gene symbol and NCBI gene ID\n",
    "ID_map = pd.read_csv(\"/Users/guangrong/Documents/GitHub_project/fastqpi_BigGIM/KGs/ID_convert/Table_ID_convert_9.19.2021.tsv\", sep = '\\t')\n",
    "dic_ID_symbol2ncbi ={}\n",
    "import math\n",
    "ID_map.astype(str)\n",
    "for i in range(0,ID_map.shape[0]):\n",
    "    symbol = ID_map['Approved_symbol'][i]\n",
    "    \n",
    "    ncbi_id = ID_map['NCBI_Gene_ID'][i]\n",
    "    ensemble_id = ID_map['Ensembl_gene_ID'][i]\n",
    "    if math.isnan(ncbi_id) == False :\n",
    "        ncbi_id = int(ncbi_id)\n",
    "        if symbol not in dic_ID_symbol2ncbi:\n",
    "            dic_ID_symbol2ncbi[symbol] = ncbi_id\n",
    "        else:\n",
    "            print(symbol)\n",
    "for i in range(0,ID_map.shape[0]):\n",
    "    alias_list = str(ID_map[\"Alias_symbols\"][i]).split(',')\n",
    "    ncbi_id = ID_map['NCBI_Gene_ID'][i]\n",
    "    ensemble_id = ID_map['Ensembl_gene_ID'][i]\n",
    "    if len(alias_list) > 0:\n",
    "        for alias in alias_list:\n",
    "            if alias not in dic_ID_symbol2ncbi:\n",
    "                dic_ID_symbol2ncbi[alias] = ncbi_id\n",
    "dic_ID_symbol2ncbi['MEN1'] = 4221\n",
    "dic_ID_symbol2ncbi['TSC1']= 7248\n",
    "dic_ID_symbol2ncbi['WHSC1'] = 7468\n",
    "dic_ID_symbol2ncbi['MET'] = 4233\n",
    "dic_ID_symbol2ncbi['HIST1H1E'] = 3008\n",
    "dic_ID_symbol2ncbi['FLNA'] = 2316\n",
    "dic_ID_symbol2ncbi[\"FAM46D\"] = 169966\n",
    "dic_ID_symbol2ncbi[\"HIST1H1C\"] = 3006"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "656aea73",
   "metadata": {},
   "outputs": [],
   "source": [
    "driver_tcga_freq = pd.read_csv(\"../Input_Table/TCGA_driver_mut_freq.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "36974c88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_name</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>subject_id_prefix</th>\n",
       "      <th>object_name</th>\n",
       "      <th>object_id</th>\n",
       "      <th>object_category</th>\n",
       "      <th>object_id_prefix</th>\n",
       "      <th>predicate</th>\n",
       "      <th>frequency_qualifier</th>\n",
       "      <th>subject_aspect_qualifier</th>\n",
       "      <th>has_count</th>\n",
       "      <th>supporting_study_cohort</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene</td>\n",
       "      <td>TCGA-ACC</td>\n",
       "      <td>8734</td>\n",
       "      <td>Disease</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>0.184783</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>17</td>\n",
       "      <td>TCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>1499</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene</td>\n",
       "      <td>TCGA-ACC</td>\n",
       "      <td>8734</td>\n",
       "      <td>Disease</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>14</td>\n",
       "      <td>TCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEN1</td>\n",
       "      <td>4221</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene</td>\n",
       "      <td>TCGA-ACC</td>\n",
       "      <td>8734</td>\n",
       "      <td>Disease</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>0.076087</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>7</td>\n",
       "      <td>TCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>APOB</td>\n",
       "      <td>338</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene</td>\n",
       "      <td>TCGA-ACC</td>\n",
       "      <td>8734</td>\n",
       "      <td>Disease</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>0.065217</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>6</td>\n",
       "      <td>TCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRKAR1A</td>\n",
       "      <td>5573</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene</td>\n",
       "      <td>TCGA-ACC</td>\n",
       "      <td>8734</td>\n",
       "      <td>Disease</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>0.054348</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>5</td>\n",
       "      <td>TCGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_name  subject_id subject_category subject_id_prefix object_name  \\\n",
       "0         TP53        7157             Gene          NCBIGene    TCGA-ACC   \n",
       "1       CTNNB1        1499             Gene          NCBIGene    TCGA-ACC   \n",
       "2         MEN1        4221             Gene          NCBIGene    TCGA-ACC   \n",
       "3         APOB         338             Gene          NCBIGene    TCGA-ACC   \n",
       "4      PRKAR1A        5573             Gene          NCBIGene    TCGA-ACC   \n",
       "\n",
       "   object_id object_category object_id_prefix                       predicate  \\\n",
       "0       8734         Disease            MONDO  gene_associated_with_condition   \n",
       "1       8734         Disease            MONDO  gene_associated_with_condition   \n",
       "2       8734         Disease            MONDO  gene_associated_with_condition   \n",
       "3       8734         Disease            MONDO  gene_associated_with_condition   \n",
       "4       8734         Disease            MONDO  gene_associated_with_condition   \n",
       "\n",
       "   frequency_qualifier subject_aspect_qualifier  has_count  \\\n",
       "0             0.184783                 Mutation         17   \n",
       "1             0.152174                 Mutation         14   \n",
       "2             0.076087                 Mutation          7   \n",
       "3             0.065217                 Mutation          6   \n",
       "4             0.054348                 Mutation          5   \n",
       "\n",
       "  supporting_study_cohort  \n",
       "0                    TCGA  \n",
       "1                    TCGA  \n",
       "2                    TCGA  \n",
       "3                    TCGA  \n",
       "4                    TCGA  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "driver_tcga_freq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7db36d78",
   "metadata": {},
   "outputs": [],
   "source": [
    "subject_name = [\"FANCA\", \"FANCC\", \"BRCA2\", \"FANCD2\",\n",
    "                \"FANCE\", \"FANCF\", \"FANCG\", \"FANCI\", \"BRIP1\",\n",
    "                \"FANCL\", \"FANCM\", \"PALB2\", \"RAD51C\", \"SLX4\", \n",
    "                \"ERCC4\", \"BRCA1\", \"UBE2T\", \"XRCC2\", \n",
    "                \"MAD2L2\", \"RFWD3\", \"RAD51\"]\n",
    "subject_id_list = []\n",
    "for name in subject_name:\n",
    "    if name in dic_ID_symbol2ncbi:\n",
    "        subject_id_list.append(dic_ID_symbol2ncbi[name])\n",
    "    else:\n",
    "        print(name)\n",
    "\n",
    "subject_id_prefixes = ['NCBIGene'] * len(subject_name)\n",
    "subject_category = [\"Gene\"] * len(subject_name)\n",
    "\n",
    "object_name = [\"Fanconi anemia\"] * len(subject_name)\n",
    "object_id = [\"MONDO:0019391\"] * len(subject_name)\n",
    "object_id_prefix = [\"MONDO\"] * len(subject_name)\n",
    "object_category = [\"Disease\"] * len(subject_name)\n",
    "\n",
    "predicate = [\"gene_associated_with_condition\"] * len(subject_name)\n",
    "subject_aspect_qualifier = ['Genetic variants'] * len(subject_name)\n",
    "knowledge_source = [\"PMID: 35854323\"] * len(subject_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "01ddbdcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "FA_mut_grap = pd.DataFrame()\n",
    "FA_mut_grap['subject_name'] = subject_name\n",
    "FA_mut_grap['subject_id_prefix'] =subject_id_prefixes\n",
    "FA_mut_grap['subject_id'] = subject_id_list\n",
    "FA_mut_grap['subject_category'] =subject_category\n",
    "\n",
    "FA_mut_grap['object_name'] = object_name\n",
    "FA_mut_grap['object_id'] = object_id\n",
    "FA_mut_grap['object_id_prefix'] = object_id_prefix\n",
    "FA_mut_grap['object_category'] = object_category\n",
    "\n",
    "FA_mut_grap['predicate'] = predicate\n",
    "FA_mut_grap['subject_aspect_qualifier'] = subject_aspect_qualifier\n",
    "FA_mut_grap['knowledge_source'] =knowledge_source"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5af90fe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "FA_mut_grap.to_csv(\"../Input_Table/FA_mut.csv\", index = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6e2e5f4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
